blockstore/splitstore/markset_badger.go

```diff
@@ -188,6 +136,140 @@ func (s *BadgerMarkSet) Has(c cid.Cid) (bool, error) {
 	}
 }
 
+func (s *BadgerMarkSet) Visit(c cid.Cid) (bool, error) {
+	key := c.Hash()
+	pendKey := string(key)
+
+	checkPending := func() (bool, error) {
+		if s.pend == nil {
+			return false, errMarkSetClosed
+		}
+
+		if _, ok := s.pend[pendKey]; ok {
+			return false, nil
+		}
+
+		for _, wr := range s.writing {
+			if _, ok := wr[pendKey]; ok {
+				return false, nil
+			}
+		}
+
+		return true, nil
+	}
+
+	s.mx.RLock()
+
+	visit, err := checkPending()
+	if !visit {
+		s.mx.RUnlock()
+		return visit, err
+	}
+
+	err = s.db.View(func(txn *badger.Txn) error {
+		_, err := txn.Get(key)
+		return err
+	})
+
+	switch err {
+	case nil:
+		s.mx.RUnlock()
+		return false, nil
+
+	case badger.ErrKeyNotFound:
+		// need to upgrade the lock to exclusive in order to write; take the version count to see
+		// if there was another write while we were upgrading
+		version := s.version
+		s.mx.RUnlock()
+
+		s.mx.Lock()
+
+		// we have to do the check dance again
+		visit, err := checkPending()
+		if !visit {
+			s.mx.Unlock()
+			return visit, err
+		}
+
+		if version != s.version {
+			// something was written to the db, we need to check it
+			err = s.db.View(func(txn *badger.Txn) error {
+				_, err := txn.Get(key)
+				return err
+			})
+
+			switch err {
+			case nil:
+				s.mx.Unlock()
+				return false, nil
+
+			case badger.ErrKeyNotFound:
+
+			default:
+				s.mx.Unlock()
+				return false, xerrors.Errorf("error checking badger markset: %w", err)
+			}
+		}
+
+		s.pend[pendKey] = struct{}{}
+		if len(s.pend) < badgerMarkSetBatchSize {
+			s.mx.Unlock()
+			return true, nil
+		}
+
+		if err := s.putPending(); err != nil {
+			return false, err
+		}
+
+		return true, nil
+
+	default:
+		s.mx.RUnlock()
+		return false, xerrors.Errorf("error checking badger markset: %w", err)
+	}
+}
+
+// writer holds the lock
+func (s *BadgerMarkSet) putPending() error {
+	pend := s.pend
+	seqno := s.seqno
+	s.seqno++
+	s.writing[seqno] = pend
```

Comment:
So, with a background worker, we can get rid of this and just have a single "writing" and "pending" set. When the background worker is done flushing the last set, it would check to see if it should flush the next one (e.g., use a channel as a flag).

That _should_ help simplify things a bit.

Reply:
The disadvantage of the background worker is that we can do only one write at the time -- this can do multiple concurrent writes, which is strictly better i think.

Reply:
It also loses the timely error return property.

I guess we could have a worker that spawns goroutines to do writes, and somehow bubble up the error to be checked in the next operation, but this gets more complicated than I'd like it to be.

Reply:
I _think_ badger will handle write parallelism under the covers.

Reply:
hrm not sure -- I prefer to be explicitly parallel and also keep the ability to return an error at some writer so that we can abort.

On the same time, I don't feel terribly strongly about this and we could use a background worker who spawns writers and bubbles up the error in some way.